components/timezone/src/ids.rs

```diff
@@ -0,0 +1,371 @@
+/// );
+///
+/// // We can recover the canonical name from the mapper:
+/// assert_eq!(
+///     mapper
+///         .canonicalize_iana("Australia/Victoria")
+///         .unwrap()
+///         .string,
+///     "Australia/Melbourne"
+/// );
+/// ```
+#[derive(Debug, Clone)]
+pub struct TimeZoneIdMapper {
+    data: DataPayload<IanaToBcp47MapV2Marker>,
+}
+
+impl TimeZoneIdMapper {
+    /// Creates a new [`TimeZoneIdMapper`] using compiled data.
+    ///
+    /// See [`TimeZoneIdMapper`] for an example.
+    ///
+    /// ✨ *Enabled with the `compiled_data` Cargo feature.*
+    ///
+    /// [📚 Help choosing a constructor](icu_provider::constructors)
+    #[cfg(feature = "compiled_data")]
+    pub fn new() -> Self {
+        Self {
+            data: DataPayload::from_static_ref(
+                crate::provider::Baked::SINGLETON_TIME_ZONE_IANA_TO_BCP47_V2,
+            ),
+        }
+    }
+
+    icu_provider::gen_any_buffer_data_constructors!(locale: skip, options: skip, error: DataError,
+        #[cfg(skip)]
+        functions: [
+            new,
+            try_new_with_any_provider,
+            try_new_with_buffer_provider,
+            try_new_unstable,
+            Self,
+        ]
+    );
+
+    #[doc = icu_provider::gen_any_buffer_unstable_docs!(UNSTABLE, Self::new)]
+    pub fn try_new_unstable<P>(provider: &P) -> Result<Self, DataError>
+    where
+        P: DataProvider<IanaToBcp47MapV2Marker> + ?Sized,
+    {
+        let data = provider.load(Default::default())?.take_payload()?;
+        Ok(Self { data })
+    }
+
+    /// Returns a borrowed version of the mapper that can be queried.
+    ///
+    /// This avoids a small potential cost of reading the data pointer.
+    pub fn as_borrowed(&self) -> TimeZoneIdMapperBorrowed {
+        TimeZoneIdMapperBorrowed {
+            data: self.data.get(),
+        }
+    }
+}
+
+/// A borrowed wrapper around the time zone ID mapper, returned by
+/// [`TimeZoneIdMapper::as_borrowed()`]. More efficient to query.
+#[derive(Debug, Copy, Clone)]
+pub struct TimeZoneIdMapperBorrowed<'a> {
+    data: &'a IanaToBcp47MapV2<'a>,
+}
+
+impl<'a> TimeZoneIdMapperBorrowed<'a> {
+    /// Gets the BCP-47 time zone ID from an IANA time zone ID
+    /// with a case-insensitive lookup.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_timezone::TimeZoneBcp47Id;
+    /// use icu_timezone::TimeZoneIdMapper;
+    /// use tinystr::tinystr;
+    ///
+    /// let mapper = TimeZoneIdMapper::new();
+    /// let mapper = mapper.as_borrowed();
+    ///
+    /// assert_eq!(
+    ///     mapper.iana_to_bcp47("AMERICA/chicago"),
+    ///     Some(TimeZoneBcp47Id(tinystr!(8, "uschi")))
+    /// );
+    /// ```
+    pub fn iana_to_bcp47(&self, iana_id: &str) -> Option<TimeZoneBcp47Id> {
+        self.iana_lookup_quick(iana_id)
+            .and_then(|trie_value| self.data.bcp47_ids.get(trie_value.index()))
+    }
+
+    /// Normalizes the syntax of an IANA time zone ID.
```

Comment:
It needs to be data driven: `America/Argentina/ComodRivadavia`

The previous PR that this replaces had a fallback algorithm if no data was found. In this new PR I defer that problem to the user. I think it's nice for it to be people data driven and then we can return the BCP ID as well and save another pass through the trie.

Reply:
> It needs to be data driven: America/Argentina/ComodRivadavia

Please elaborate?

Reply:
There are about 20 IANA names whose normalization does not follow the `to_ypotryll_case` function in #4548. `America/Argentina/ComodRivadavia` would be incorrectly normalized to `America/Argentina/Comodrivadavia` if performed without data.

I'm potentially open to the idea of unknown IDs defaulting to `to_ypotryll_case` but as noted above I don't think it's necessarily the library's job to do that.

Reply:
So what's the point of distinguishing between normalisation and canonicalisation?

Reply:
Canonicalization can change the city, too. `America/Argentina/ComodRivadavia` canonicalizes to `America/Argentina/Catamarca`.

ECMAScript Temporal cares about this distinction because normalization is fixed but canonicalization can change over time.

Reply:
Note that many of these are deprecated aliases but there are at least a few that are canonical.

- Africa/Dar_es_Salaam
- America/Port-au-Prince
- America/Port_of_Spain
- Antarctica/DumontDUrville
- Antarctica/McMurdo
- All the Etc/ time zones
- Europe/Isle_of_Man

Reply:
Please document why normalization needs data, and that usually the user probably wants canonicalization.

Reply:
I added a section about the difference between the operations.